### PR TITLE
Feat/mgv order gas tests

### PR DESCRIPTION
**Pull request by jkrivine**
This PR does the following:
* make consistent and comprehensive tests for `MangroveOrder`, with tests for:
   * Pure taker side (running m.o with MangroveOrder)
   * Maker side (posting resting order)
   * manager side (resting order management)
   * gas requirements
* it establishes the following values (use `forge test --mc MangroveOrder --mt mockup_offerLogic_gas_cost -vv` to check):
   * worst case gas requirement for resting order's logic is ~ 125K with ~80K for execution and 45K for posthook.
   * note gas requirement should not count router gas cost which is roughly 70K (us `forge test --mc MangroveOrder --mt mockup_routing_gas_cost -vv`)
   * so overall this suggests to set `gasreq` for MangroveOrder at 60K (if Router's gasreq remains at 70K)

### CODE REVIEW

File: .solhint.json [hunk 1]
```diff
@@ -6,12 +6,23 @@
...
+    "reason-string": [
+      "warn",
+      {
+        "maxLength": 40
+      }
+    ],
     "not-rely-on-time": "off",
-    "avoid-low-level-calls": "off"
+    "avoid-low-level-calls": "off",
+    "no-console": "off"
```
Comment (peterMangrove) [inline .solhint.json]: Why add this?

File: test/lib/MangroveTest.sol [hunk 2]
```diff
@@ -145,19 +145,22 @@ contract MangroveTest is Test2, HasMgvEvents {
...
+        mgv.offerInfo($out, $in, offerId);
       console.log(
         string.concat(
-          unicode"│ ",
-          string.concat(offerId < 9 ? " " : "", vm.toString(offerId)), // breaks on id>99
-          unicode" ┆ ",
-          string.concat(toUnit(ofr.wants, req_tk.decimals()), " ", req_tk.symbol()),
-          "  /  ",
-          string.concat(toUnit(ofr.gives, ofr_tk.decimals()), " ", ofr_tk.symbol())
+          string.concat(
```
Comment (peterMangrove) [inline test/lib/MangroveTest.sol]: Why add an extra `string.concat`?

File: test/lib/MangroveTest.sol [hunk 3]
```diff
@@ -408,4 +410,37 @@ contract MangroveTest is Test2, HasMgvEvents {
...
+      fold--;
+    }
+  }
+
+  /// duplicates `fold` times all offers in the `outbound, inbound` list from id `fromId` and for `lenght` offers.
+  function densify(address outbound, address inbound, uint fromId, uint length, uint fold, address caller) internal {
+    while (length > 0 && fromId != 0) {
+      MgvStructs.OfferPacked offer = mgv.offers(outbound, inbound, fromId);
+      MgvStructs.OfferDetailPacked detail = mgv.offerDetails(outbound, inbound, fromId);
+      densify(outbound, inbound, offer.wants(), offer.gives(), detail.gasreq(), fromId, fold, caller);
```
Comment (peterMangrove) [inline test/lib/MangroveTest.sol]: For me, it is a bit confusing that you have 2 `densify` methods, that does different things. Maybe call one of them `innerDensify`, just to clarify the difference

**Review by peterMangrove — COMMENTED** [review]
Made a few comments. Did not go through all the new test and test changes in detail. But otherwise it looks good.

File: test/strategies/MgvOrder.t.sol [hunk 4]
```diff
@@ -272,481 +333,489 @@ contract MangroveOrder_Test is MangroveTest {
...
-    vm.expectRevert("mgvOrder/refundFail");
-    vm.prank($(buy_taker));
-    // complete fill will not lead to a resting order
-    mgo.take{value: 0.1 ether}(buyOrder);
+    // mock up mgv taker transfer to maker contract
+    base.transfer($(mgo), 0.5 ether);
+    sellOrder.offerId = cold_buyResult.offerId;
+    vm.prank($(mgv));
+    _gas();
+    mgo.makerExecute(sellOrder);
```
Comment (lnist) [inline test/strategies/MgvOrder.t.sol:801]: I read your tests correctly it seems we are including the cost of reading offer information in the estimate - but I would expect offer information to already have been read by the core protocol when it invokes makerExecute?

File: test/strategies/MgvOrder.t.sol [hunk 5]
```diff
@@ -272,481 +333,489 @@ contract MangroveOrder_Test is MangroveTest {
...
+  function test_mockup_routing_gas_cost() public {
+    SimpleRouter router = SimpleRouter(address(mgo.router()));
+    // making quote balance hot to mock taker's transfer
+    quote.transfer($(mgo), 1);
 
-    vm.startPrank($(buy_taker));
-    TransferLib.approveToken(quote, $(mgo.router()), type(uint).max);
-    vm.stopPrank();
+    vm.prank($(mgo));
+    uint g = gasleft();
```
Comment (lnist) [inline test/strategies/MgvOrder.t.sol:769]: why gaslesft() over _gas here?
Also should this be in a router test and not in order test? Its ok for now :)

File: test/strategies/MgvOrder.t.sol (hunk 4 above)
Comment (jkrivine) [reply test/strategies/MgvOrder.t.sol:801]: Right I updated this

File: script/strategies/mangroveOrder/MangroveOrderDeployer.s.sol [hunk 6]
```diff
@@ -28,11 +28,13 @@ contract MangroveOrderDeployer is Deployer {
...
     MangroveOrder mgvOrder;
+    // use 30K gasreq, this will be addeed to the SimpleRouter gasreq of 70K.
+    // tests show that MangroveOrder requires 65K under normal circumstances.
     broadcast();
     if (forMultisig) {
-      mgvOrder = new MangroveOrder{salt:salt}(mgv, admin, 30_000);
+      mgvOrder = new MangroveOrder{salt:salt}(mgv, admin, 50_000);
     } else {
-      mgvOrder = new MangroveOrder(mgv, admin, 30_000);
+      mgvOrder = new MangroveOrder(mgv, admin, 50_000);
```
Comment (peterMangrove) [inline script/strategies/mangroveOrder/MangroveOrderDeployer.s.sol]: Either the comment is wrong or the amount given as the actual gasReq is wrong. The comment says 30K, but you use 50K?

**Review by peterMangrove — CHANGES_REQUESTED** [review]

File: test/lib/MangroveTest.sol (hunk 3 above)
Comment (jkrivine) [reply test/lib/MangroveTest.sol]: Called the second `densifyRange`

**Review by peterMangrove — APPROVED** [review]
lgtm, expect for the failing CI 😅

**Pull request merged**